### PR TITLE
Closes #2468 Implement derive_vars_crit_flag()

### LINKED ISSUE #2468: Feature Request: Implement `derive_vars_crit_flag()`
### Feature Idea

Implement `derive_vars_crit_flag()` which derives the variables `CRITy`, `CRITyFL`, and `CRITyFLN` (for a specific value of `y`). The derivation of these variables is very simple. However, there are some requirements which many programmers and specs authors are not aware of. For example, that the derivation must rely on variables from the current records only and the relation of the values of `CRITy` and `CRITyFL` (e.g., if `CRITyFL` is set to `"Y"` or `"N"`, `CRITy` must be populated for all records with the same value).

The function should cover all use cases of the `CRIT` variables. I.e., if a `CRIT` variable can not be derived with the function, the derivation is not ADaM compliant. 

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

```
adxx <- adxx %>% derive_vars_crit_flag(
  crit_nr = 1,
  condition = AVAL > 50,
  description = paste(PARAM, "> 50"),   # CRITy is set to the specified value
  values_yn = FALSE,                    # should values "Y", "N", and NA be used?
  create_numeric_flag = TRUE            # should CRITyFLN be created?
)
```

### PR DESCRIPTION
**Pull request by bundfussr**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF version of it to the same folder.
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers). A Developer Notes section is available in `NEWS.md` for tracking developer-facing issues.  
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/derive_vars_crit_flag.R [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+#'   crit_nr = 2,
+#'   condition = AVAL > 50,
+#'   description = paste(PARAMCD, "> 50"),
+#'   values_yn = TRUE,
+#'   create_numeric_flag = TRUE
+#' )
+derive_vars_crit_flag <- function(dataset,
+                                  crit_nr = 1,
+                                  condition,
+                                  description,
```
Comment (bms63) [inline R/derive_vars_crit_flag.R:93]: Do we use description any where else?  Wondering if there is another argument we could use that is similar to other functions??

**Comment by bms63** [comment]
Sorry I still want to review more thoroughly.  Can we not merge until I complete please.

**Comment by bundfussr** [comment]
> LGTM - only comment is, can we showcase this new function in a vignette? What about [here](https://pharmaverse.github.io/admiral/dev/articles/bds_finding.html#cat) or right after?

Yes, good idea. I'll add an example there.

**Comment by bms63** [comment]
What do you think about a MCRIT flag for multiple responses?  Worth to make a related function?

**Comment by bms63** [comment]
Worth to give more reasoning to a user on why these flags might be useful? 
image

**Comment by bms63** [comment]
oohhh could we have a similar example with the AND - this is from ADaM IG 
image

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ Please check off each taskbox as an acknowledgment that you completed the task o
...
 - [ ] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
 - [ ] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
-- [ ] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF version of it to the same folder.
+- [ ] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF and a PNG version of it to the same folder. (The PNG version can be created by taking a screenshot of the PDF version.)
```
Comment (bms63) [inline .github/pull_request_template.md:9]: Nice addition to the checklist!

File: tests/testthat/test-derive_vars_crit_flag.R [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      description = paste(PARAMCD, "> 40"),
+      values_yn = TRUE,
+      create_numeric_flag = TRUE
+    ),
+    keys = "AVAL"
+  )
+})
+
+## Test 4: error if invalid condition ----
+test_that("derive_vars_crit_flag Test 4: error if invalid condition", {
```
Comment (bms63) [inline tests/testthat/test-derive_vars_crit_flag.R:68]: Can we specify that the variable is missing in the input dataset

File: tests/testthat/test-derive_vars_crit_flag.R [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      input,
+      condition = AVAL > 3 * ANRHI,
+      description = "> 3ULN"
+    ),
+    error = TRUE
+  )
+})
+
+## Test 5: error if invalid description ----
+test_that("derive_vars_crit_flag Test 5: error if invalid description", {
```
Comment (bms63) [inline tests/testthat/test-derive_vars_crit_flag.R:87]: Can we specify that the variable PARAMCD is not in the dataset

**Review by bms63 — CHANGES_REQUESTED** [review]
I think I just have some concerns around testing and the AND condition example.  Everything else is looking super neat and awesome!

**Comment by bundfussr** [comment]
> LGTM - only comment is, can we showcase this new function in a vignette? What about [here](https://pharmaverse.github.io/admiral/dev/articles/bds_finding.html#cat) or right after?

@manciniedoardo , I have added an example. It is also a nice use case of our higher order functions.

**Comment by bundfussr** [comment]
> oohhh could we have a similar example with the AND condition - this is from ADaM IG image

@bms63 , I have added this example to the bds_finding vignette.

File: tests/testthat/test-derive_vars_crit_flag.R (hunk 3 above)
Comment (bundfussr) [reply tests/testthat/test-derive_vars_crit_flag.R]: Done

File: tests/testthat/test-derive_vars_crit_flag.R (hunk 4 above)
Comment (bundfussr) [reply tests/testthat/test-derive_vars_crit_flag.R]: Done

**Comment by bundfussr** [comment]
> Worth to give more reasoning to a user on why these flags might be useful? image

@bms63 , I've added it to the description of the function.

File: tests/testthat/test-derive_vars_crit_flag.R
[the remark this answers is not in the document]
Comment (bundfussr) [reply tests/testthat/test-derive_vars_crit_flag.R:1]: I'm not sure if we need additional checks for the condition. I think the important point is that the condition is evaluated correctly. I hope this is covered by the existing tests.

I am aware only of the restriction that the condition must use variables/value of the same record. Apart from that you could use arbitrary complex conditions. For example `((AVAL < 50 & CHG < -10) | (AVAL > 150 & CHG > 10)) & (0 < AVISITN & (AVISITN < 12 | AVISIT == "UNSCHEDULED))` is permitted in my opinion.

File: R/derive_vars_crit_flag.R (hunk 1 above)
Comment (bms63) [reply R/derive_vars_crit_flag.R:95]: I don't - just wanted to be consistent.  Thanks for looking into other examples

**Review by bms63 — APPROVED** [review]
The first function for 1.2!! Way to go @bundfussr 

@manciniedoardo we should start a skeleton blog for 1.2 and add this in.

**Comment by bundfussr** [comment]
> What do you think about a MCRIT flag for multiple responses? Worth to make a related function?

I wouldn't implement an `MCRIT` specific function because there are almost no restrictions on the values of the `MCRIT*` variables. However, a function for creating variable pairs like `MCRITyFL`/`MCRITyFN` in one go would be useful. See https://github.com/pharmaverse/admiral/issues/2480.

**Pull request merged**